functions/src/winloss.ts

```diff
@@ -0,0 +1,125 @@
+  return db.runTransaction(async (transaction) => {
+    const lobby = await transaction.get(lobbyDocRef);
+    const lobbyData = lobby.data();
+
+    if (lobbyData === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
+    const privatePlayerCollection = getPrivatePlayerCollection(lobbyDocRef);
+    const { players, uids } = lobbyData;
+    await Promise.all(
+      lobbyData.uids.map(async (uid) => {
+        const privatePlayerDocRef = privatePlayerCollection.doc(uid);
+        const privatePlayerDoc = await transaction.get(privatePlayerDocRef);
+        const privatePlayerData = privatePlayerDoc.data();
+
+        if (!privatePlayerData) {
+          throw new Error(`Private player not found for uid ${uid}`);
+        }
+        const playerId: string = uid;
+        players[uids.indexOf(playerId)].role = privatePlayerData.role;
+      })
+    );
+
+    let catfishCount = 0;
+    for (let i = 0; i < uids.length; i++) {
+      // check the alive players
+      if (lobbyData.players[i].alive == true) {
+        if (lobbyData.players[i].role == "CATFISH") {
+          catfishCount++;
+        }
+      }
+    }
+    if (catfishCount == 0) {
+      lobbyData.winner = "CAT";
+    } else {
+      lobbyData.winner = "CATFISH";
+    }
+    transaction.set(lobbyDocRef, lobbyData);
+  });
+}
+
+// when returning to the lobby, make every player alive again and delete the role, winner, and the private player collection
+export function resetLobby(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+  return db.runTransaction(async (transaction) => {
+    const lobby = await transaction.get(lobbyDocRef);
+    const lobbyData = lobby.data();
+
+    if (lobby === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
+
+    if (lobbyData === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
+
+    // loop through each player and set their alive status to true and remove the role property from the player
+    let { players } = lobbyData;
+    players = players.map(({ role, ...rest }) => {
+      return { ...rest, alive: true };
+    });
+
+    transaction.update(lobbyDocRef, { state: "WAIT", players, winner: firestore.FieldValue.delete() });
+  });
+}
+
+export async function deletePrivatePlayers(
+  lobbyDocRef: firestore.DocumentReference<Lobby>,
+  privatePlayerCollectionRef: firestore.CollectionReference<PrivatePlayer>
+) {
+  const lobby = await lobbyDocRef.get();
+  if (lobby.exists === false) {
+    throw new Error("Lobby does not exist!");
+  }
+  // create a WriteBatch
+  const batch = db.batch();
+  // loop through each doc in the private player collection via uid and delete them one-by-one
+  const { uids } = lobby.data() as Lobby;
+  for (let i = 0; i <= uids.length; i++) {
+    const playerDoc = privatePlayerCollectionRef.doc(uids[i]);
+    // delete each private player document with the WriteBatch
+    batch.delete(playerDoc);
+    batch.commit();
```

Comment:
the commit should be outside the loop and should be awaited

Reply:
I originally had the commit outside the loop, but the delete wasn't being committed

Reply:
> I originally had the commit outside the loop, but the delete wasn't being committed

did you forget to await the commit?

Reply:
no, I'm testing it right now too and it's not committing the deletes

Reply:
nvm I had `i <= uids.length` in my for loop and had to remove the `=` lmao